w3af/plugins/infrastructure/webapp_detection.py

```diff
@@ -0,0 +1,139 @@
+"""
+webapp_detection.py
+
+Copyright 2015 Piotr Lizonczyk
+
+This file is part of w3af, http://w3af.org/ .
+
+w3af is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation version 2 of the License.
+
+w3af is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with w3af; if not, write to the Free Software
+Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+
+"""
+from wad.detection import Detector, TIMEOUT
+from wad.output import HumanReadableOutput
+
+import w3af.core.controllers.output_manager as om
+import w3af.core.data.kb.knowledge_base as kb
+
+from w3af.core.controllers.plugins.infrastructure_plugin import InfrastructurePlugin
+from w3af.core.controllers.exceptions import RunOnce
+from w3af.core.controllers.misc.decorators import runonce
+from w3af.core.data.kb.info import Info
+from w3af.core.data.options.opt_factory import opt_factory
+from w3af.core.data.options.option_list import OptionList
+
+
+class w3afDetector(Detector):
+    """
+    This class overrides some getter functions in original Detector class
+    in order to use w3af's ExtendedUrllib.
+    """
+    def __init__(self, uri_opener):
+        super(w3afDetector, self).__init__()
+        self._uri_opener = uri_opener
+
+    def get_page(self, url, timeout=TIMEOUT):
+        return self._uri_opener.GET(url, cache=True, timeout=timeout)
+
+    def get_new_url(self, page):
+        return page.get_url().url_string
+
+    def get_content(self, page, url):
+        return page.get_body()
+
+
+class webapp_detection(InfrastructurePlugin):
+    """
+    Identify technologies used by website based on HTTP response and HTML content
+
+    This plugin is a wrapper of CERN's CERT team's Web application detection (WAD) tool.
+
+    :author: Piotr Lizonczyk (piotr.lizonczyk@gmail.com)
+    """
+    def __init__(self):
+        InfrastructurePlugin.__init__(self)
+
+        # User configured
+        self._timeout = 5
+
+    @runonce(exc_class=RunOnce)
```

Comment:
The `RunOnce` is something I would also use, but I was wondering... is this the best we can do? Let me give you a couple of examples:

 * The first fuzzable request that arrives to this plugin is usually http://foo.com/ , in some cases that might just be a redirect to something else like http://foo.com/home/ . w3af doesn't automatically follow redirects, so wad would analyze an empty result

 * http://foo.com/ is just a simple page with links to the different application sections, / doesn't use any JS frameworks nor CMS/etc, but the deeply linked apps do.

In both those cases WAD could find stuff and we're not invoking it to do so.

What if we let w3af call WAD several times until no more new information is collected and then we raise `NoMoreCalls`? Broken pseudo-code:

```python
url_knowledge = self.analyze_url(url)
if self.is_new_knowledge(url_knowledge):
    self.wad_knowledge.merge(url_knowledge)
    self.no_new_knowledge_count = 0
else:
    self.no_new_knowledge_count += 1
    if self.no_new_knowledge_count == 10:
        raise NoMoreCalls
```

I randomly chose `10`, you can change that to any value based on your previous experience with wad